### ISSUE TITLE
[rv_dm] CDC and RDC linting and closure

### ISSUE THREAD

**Issue by andreaskurth** [lowRISC/opentitan#15107]
Clock domain crossing (CDC) and reset domain crossing (RDC) linting and closure happens at the top level. As `rv_dm` is quite involved w.r.t. those two topics, this issue exists to explicitly raise awareness and track progress of CDC and RDC closure. It can be closed once both closures have been reached.

## Notes
- `dmi_cdc` contains a "TODO" comment saying that the JTAG reset must be made clean for synthesis.

**Comment by eunchan** [comment]
Current `prim_fifo_async` has one bit pointers inside if `Depth` is 1.

If JTAG TRST is asserted when the pointer is 1, then there's chance the metastability issue occurs on the core clock side.

One way to resolve this is to make the async FIFO synchronous reset. It is not as of now (async assert / sync de-assert for JTAG TRST) in `dmi_cdc`.

Either we can waive the scenario as asserting TRST_N during the JTAG debugging is rare in my opinion, or making synchronous reset in `dmi_cdc` would be solutions.

CC: @tjaychen @msfschaffner

**Comment by eunchan** [comment]
If I add constraints to the req/rsp_valid signals on dmi_cdc, all RDC errors are gone.

**Comment by msfschaffner** [comment]
I believe our assumption was that the debugger can sequence TRST correctly with any JTAG transactions that may be in flight. Hence waiving seems ok. Corresponding PR where this assumption has been stated: 
https://github.com/lowRISC/opentitan/pull/12613#issue-1233274955

**Comment by eunchan** [comment]
Hum, I think even there's no outstanding transaction in `dmi_cdc`, still issue may occur, as the `fifo_?ptr` inside `prim_fifo_async` could be any values (0 or 1) in idle state.

So, the issue may occur when `fifo_?ptr` are 1. Then, TRST_N may reset them to 0 (asynchronously). It creates the metastability issue on the core clock side as the receiver (consumer of the DMI req) is not reset by TRST_N.

So, make TRST_N synchronously asserting on the core clock side resolves the issue in my opinion.

**Comment by eunchan** [comment]
I created the other method, which I assume is wrong, in https://github.com/lowRISC/opentitan/pull/16449

**Comment by msfschaffner** [comment]
Ah I see what you are saying. Yeah this could become a problem.

Even if we made the reset synchronous on the core side, wouldn't we still have a potential reconvergent path problem?
I.e., the synchronized reset on the core side would have to be asserted before the JTAG side pointer has been synchronized to the core side - otherwise we could get a ghost transaction on the core side due to the pointers being out of sync.

**Comment by eunchan** [comment]
I am thinking this a little bit.

TL;DR is to reset the consumer side first then the producer side. So, in `dmi_cdc` TRST_N propagates to core_clk first then the reset is bace synched to TCK domain, which resets the CDC. Then we should not have any issue. Benefit is the transaction does not need any additional delays like #16483 and FIFO depth can be greater than 1 (the reset scheme can be used in other FIFOs too)

Let me draw the waveform after getting back from vacation.

CC: @tjaychen @msfschaffner

**Comment by eunchan** [comment]
Hum.. maybe no need of syncing reset from core clock side to create the TCK side reset. Just passing two posedge of TCK should be fine as no matter how fast TCK is (usually slower though), always core_clk side is reset first then the synced pointers will arrive.

As long as TRST_N holds the reset for long enough (to cover the 2 TCK + 2 core_clk), any intermediate pointer values will be ignored.

Assumption here is: core_clk side is sync assert, sync de-assert reset. TCK side reset could be async or sync assert sync de-assert.

**Comment by eunchan** [comment]
For the 2FF on both side, I think the approach is simpler and doable.
![image](https://user-images.githubusercontent.com/1192814/204618276-64011ff3-5d70-4ada-8d3a-83f4d48ff672.png)

The benefit is, inside `dmi_cdc`, we do not have to create four resets (2 for req, 2 for rsp).

**Comment by eunchan** [comment]
One thing bothers me is, as the logics inside the FIFO uses async reset, there may be a timing issue. As soon as the reset is synced, the reset affects the pointers inside the FIFO without waiting the next posedge of CLK. Then the value is propagating to the consumer logic (which is not affected by the reset). If the path is on the critical path and the synthesis tool assumes the reset sync as a relaxed reset tree, timing may not be enough in certain circumstances. I am sure we do not have that path on RV_DM, LC_CTRL, but maybe we need to have a separate sync reset FIFO module to reduce the risk.

**Comment by msfschaffner** [comment]
Yeah that is the part that I tried to avoid by making a "simpler" CDC module that does not have pointer states. Let me set up a short meeting so that we can discuss the course of action this afternoon.

**Comment by tjaychen** [comment]
sounds good, thanks for setting up the meeting.
I'll also be there in person tomorrow, if you guys want to do that. Both are okay with me.

**Comment by msfschaffner** [comment]
@eunchan I think this has been resolved now with the new CDC crossing, right?
